src/MSBuild.UnitTests/ValidateAssemblyLoadContext.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#if FEATURE_ASSEMBLYLOADCONTEXT
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+using Microsoft.Build.Utilities;
+using System.Runtime.Loader;
+
+namespace Microsoft.Build.UnitTests
+{
+    public class ValidateAssemblyLoadContext : Task
+    {
+        public override bool Execute()
+        {
+            var thisLoadContext = AssemblyLoadContext.GetLoadContext(typeof(ValidateAssemblyLoadContext).Assembly);
+
+            // Check by name because this always reports false, presumably due to ALC shenanigans:
+            // if (thisLoadContext is MSBuildLoadContext context)
```

Comment:
Did you file a follow up issue for that commented out section? That definitely seems like a bug or at least an item worthy of documentation.

Reply:
Yeah, I don't understand what's going on here.

I patched it a bit to get more details (again, I think I did this before):

```C#
Log.LogMessage(MessageImportance.High, $"thisLoadContext is MSBuildLoadContext context : {thisLoadContext is MSBuildLoadContext context}");
Log.LogMessage(MessageImportance.High, $"thisLoadContext            loaded in AssemblyLoadContext {AssemblyLoadContext.GetLoadContext(thisLoadContext.GetType().Assembly)}");
Log.LogMessage(MessageImportance.High, $"typeof(MSBuildLoadContext) loaded in AssemblyLoadContext {AssemblyLoadContext.GetLoadContext(typeof(MSBuildLoadContext).Assembly)}");
```

```
  thisLoadContext is MSBuildLoadContext context : False
  thisLoadContext            loaded in AssemblyLoadContext "Default" System.Runtime.Loader.DefaultAssemblyLoadContext #2
  typeof(MSBuildLoadContext) loaded in AssemblyLoadContext "Default" System.Runtime.Loader.DefaultAssemblyLoadContext #2
```

My theory was that they were in different ALCs but that doesn't seem to be true.

@vitek-karas do you know what might cause this?

Reply:
The different ALCs would be my theory as well - but as you mention that doesn't seem to be the case. You can try to GetType on both ends of the comparison and look it up in the debugger (make object ID) to double check. If you can prove that the Type objects are the same, but "is" doesn't work - that would be a problem.

Reply:
I feel uneasy that this isn't resolved. Please replace "shenanigans" with an explanation as what is happening. It feels like we don't even know if this is a bug or not yet.

Reply:
Thanks to Nick offline I think I now understand what's going on. Because `MSBuildLoadContext` is compiled as source into `MSBuild.exe` and `Microsoft.Build.dll`, there are two copies of it available at runtime. MSBuild's copy is exposed to the test assembly via IVT, but Microsoft.Build's is the one that is relevant for loading the task. So in fact they don't match--even though they have the same namespace names they have different assembly-qualified names.

I'll amend the comment to describe this--I don't think it's worth juggling IVT right now. Filing an issue to come back and look at that.

Reply:
What does IVT stand for?  Intermediate Value Theorem?? Interrupt Vector Table?? InternalsVisibleTo?